cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java

```diff
@@ -381,6 +383,97 @@ private PipelinePlan createPlan() {
     return planner.plan(spec);
   }
 
+  private void updateTokenWithTriggeringCompositeProperties(TriggeringScheduleInfo scheduleInfo,
```

Comment:
This function behaves exactly like `updateTokenWithTriggeringProperties`, except that it iterates through all the triggers and match the id information

Reply:
should not copy-paste code like this. Refactor things so that this method calls updateTokenWithTriggeringProperties.

While you're at it, I think it would also be a good idea to change updateTokenWithTriggeringProperties so that it is a static method that returns a Map of keys rather than mutating the WorkflowToken within the method. That way you can add unit tests that calls the method directly rather than the current tests that need to deploy and run entire pipelines

Reply:
Thanks! updated